File: android/app/src/main/java/com/openburnbar/data/square/PinnedAgentGridConfig.kt
package com.openburnbar.data.square

import com.openburnbar.data.hermes.AssistantRuntimeID
import org.json.JSONArray
import org.json.JSONObject

// MARK: - Pinned Agent Grid Config (Android parity)
//
// JSON-compatible mirror of the Swift `PinnedAgentGridConfig`. Persisted
// in SharedPreferences under `square.pinned_grid.v1`.

data class PinnedAgentGridConfig(
    val pinnedURIs: List<String> = defaultPinnedURIs(),
    val displayMode: DisplayMode = DisplayMode.COMFORTABLE,
    val lastRearrangedAtEpoch: Long? = null,
) {
    enum class DisplayMode(val token: String, val columns: Int, val rows: Int) {
        COMFORTABLE("comfortable", 4, 3),
        COMPACT("compact", 6, 2),
        ;

        companion object {
            fun fromToken(value: String?): DisplayMode = values().firstOrNull { it.token == value } ?: COMFORTABLE
        }
    }

    fun sanitized(): PinnedAgentGridConfig {
        val seen = linkedSetOf<String>()
        val deduped = pinnedURIs.filter { uri -> seen.add(uri) }
        val trimmed = deduped.take(MAX_SLOTS)
        val defaultMigrated =
            if (
                lastRearrangedAtEpoch == null &&
                trimmed == legacyDefaultPinnedURIsBeforeDroidForge()
            ) {
                defaultPinnedURIs()
            } else {
                trimmed
            }
        val nonEmpty =
            if (trimmed.isEmpty()) {
                defaultPinnedURIs().take(1)
            } else {
                defaultMigrated
            }
        return copy(pinnedURIs = nonEmpty)
    }

    fun pinning(uri: String): PinnedAgentGridConfig {
        if (uri in pinnedURIs || pinnedURIs.size >= MAX_SLOTS) return this
        return copy(
            pinnedURIs = pinnedURIs + uri,
            lastRearrangedAtEpoch = System.currentTimeMillis(),
        )
    }

    fun pinningPairedMac(uri: String): PinnedAgentGridConfig {
        if (!uri.startsWith(AgentIdentity.PAIRED_MAC_URI_PREFIX)) {
            return pinning(uri)
        }
        val reordered =
            listOf(uri) +
                pinnedURIs.filterNot {
                    it.startsWith(AgentIdentity.PAIRED_MAC_URI_PREFIX)
                }
        if (reordered.take(MAX_SLOTS) == pinnedURIs) return this
        return copy(
            pinnedURIs = reordered.take(MAX_SLOTS),
            lastRearrangedAtEpoch = System.currentTimeMillis(),
        ).sanitized()
    }

    fun hasPairedMacPin(): Boolean = pinnedURIs.any { it.startsWith(AgentIdentity.PAIRED_MAC_URI_PREFIX) }

    fun removingPairedMacPins(): PinnedAgentGridConfig = copy(
        pinnedURIs =
        pinnedURIs.filterNot {
            it.startsWith(AgentIdentity.PAIRED_MAC_URI_PREFIX)
        },
        lastRearrangedAtEpoch = System.currentTimeMillis(),
    ).sanitized()

    fun unpinning(uri: String): PinnedAgentGridConfig = copy(
        pinnedURIs = pinnedURIs.filter { it != uri },
        lastRearrangedAtEpoch = System.currentTimeMillis(),
    ).sanitized()

    fun moving(from: Int, to: Int): PinnedAgentGridConfig {
        if (pinnedURIs.isEmpty()) return this
        val safeFrom = from.coerceIn(0, pinnedURIs.size - 1)
        val safeTo = to.coerceIn(0, pinnedURIs.size)
        val mutable = pinnedURIs.toMutableList()
        val value = mutable.removeAt(safeFrom)
        mutable.add(safeTo.coerceAtMost(mutable.size), value)
        return copy(
            pinnedURIs = mutable,
            lastRearrangedAtEpoch = System.currentTimeMillis(),
        )
    }

    fun toJsonString(): String {
        val obj = JSONObject()
        obj.put("displayMode", displayMode.token)
        obj.put("pinnedURIs", JSONArray(pinnedURIs))
        if (lastRearrangedAtEpoch != null) {
            obj.put("lastRearrangedAtEpoch", lastRearrangedAtEpoch)
        }
        return obj.toString()
    }

    companion object {
        const val MAX_SLOTS = 12
        const val SHARED_PREFS_KEY = "square.pinned_grid.v1"
        const val DEFAULT_PAIRED_MAC_CONNECTION_ID = "paired-mac:default"

        val DEFAULT_PAIRED_MAC_URI: String
            get() = AgentIdentity.pairedMacURI(DEFAULT_PAIRED_MAC_CONNECTION_ID)

        fun defaultPinnedURIs(): List<String> = AssistantRuntimeID.values().map { AgentIdentity.builtInURI(it) }

        fun legacyDefaultPinnedURIsBeforeDroidForge(): List<String> = listOf(
            AssistantRuntimeID.HERMES,
            AssistantRuntimeID.PI,
            AssistantRuntimeID.CODEX,
            AssistantRuntimeID.CLAUDE,
            AssistantRuntimeID.OPEN_CLAW,
        ).map { AgentIdentity.builtInURI(it) }

        val DEFAULT = PinnedAgentGridConfig()

        fun fromJsonString(raw: String?): PinnedAgentGridConfig {
            if (raw.isNullOrBlank()) return DEFAULT
            return runCatching {
                val obj = JSONObject(raw)
                val arr = obj.optJSONArray("pinnedURIs") ?: JSONArray()
                val uris = (0 until arr.length()).map { arr.optString(it) }
                val display = DisplayMode.fromToken(obj.optString("displayMode", "comfortable"))
                val ts =
                    if (obj.has("lastRearrangedAtEpoch")) {
                        obj.optLong("lastRearrangedAtEpoch")
                    } else {
                        null
                    }
                PinnedAgentGridConfig(
                    pinnedURIs = uris,
                    displayMode = display,
                    lastRearrangedAtEpoch = ts,
                ).sanitized()
            }.getOrDefault(DEFAULT)
        }
    }
}

File: OpenBurnBarCore/Sources/OpenBurnBarCore/SharedModels/PinnedAgentGridConfig.swift
import Foundation

// MARK: - Pinned Agent Grid Config (Hermes Square §3 / §6.2)
//
// The Alipay-style 12-slot home grid at the top of the Hermes Square
// Living Inbox. Carries:
//   • The ordered list of pinned agent URIs (≤ 12).
//   • A small set of layout hints (compact vs comfortable, glyph-only vs
//     name-shown).
//
// Persisted as a JSON blob:
//   • iOS    — UserDefaults key `square.pinnedGrid.v1`
//   • Android — DataStore key `square.pinned_grid.v1`
//   • Cloud-synced via Firestore `users/{uid}/square_state/pinned_grid`
//
// Mirrors `ChatTilePreferences`'s codable patterns (deterministic key
// ordering, `sanitized()` invariant).

public struct PinnedAgentGridConfig: Codable, Sendable, Hashable {
    /// Maximum number of slots in the grid (plan-mandated: 12).
    public static let maxSlots = 12

    /// Default pinned set on first install — the five built-in runtimes in
    /// `AssistantRuntimeID.allCases` order, leaving the rest of the 12
    /// slots empty.
    public static var defaultPinnedURIs: [String] {
        AssistantRuntimeID.allCases.map(AgentIdentity.builtInURI)
    }

    /// Ordered list of pinned agent URIs.
    public var pinnedURIs: [String]

    /// Layout hints.
    public var displayMode: DisplayMode

    /// Last time the user re-arranged the grid. Used for the "you
    /// recently re-organized your dock" affordance on Mac.
    public var lastRearrangedAt: Date?

    public enum DisplayMode: String, Codable, Sendable, Hashable, CaseIterable {
        /// Glyph + abbreviated name, 4 columns × 3 rows.
        case comfortable
        /// Glyph only, 6 columns × 2 rows. More agents fit on a smaller phone.
        case compact

        public var columns: Int {
            switch self {
            case .comfortable: return 4
            case .compact:     return 6
            }
        }

        public var rows: Int {
            switch self {
            case .comfortable: return 3
            case .compact:     return 2
            }
        }
    }

    public init(
        pinnedURIs: [String] = PinnedAgentGridConfig.defaultPinnedURIs,
        displayMode: DisplayMode = .comfortable,
        lastRearrangedAt: Date? = nil
    ) {
        self.pinnedURIs = pinnedURIs
        self.displayMode = displayMode
        self.lastRearrangedAt = lastRearrangedAt
    }

    public static let `default` = PinnedAgentGridConfig()

    // MARK: - Codable

    private enum CodingKeys: String, CodingKey {
        case pinnedURIs
        case displayMode
        case lastRearrangedAt
    }

    // MARK: - Sanitisation

    /// Returns a copy with:
    ///   • duplicate URIs removed (first wins),
    ///   • list truncated to `maxSlots`,
    ///   • at least the first built-in agent pinned (so the grid is never
    ///     empty after a sanitisation pass).
    public func sanitized() -> PinnedAgentGridConfig {
        var seen: Set<String> = []
        var deduped: [String] = []
        for uri in pinnedURIs where !seen.contains(uri) {
            seen.insert(uri)
            deduped.append(uri)
        }
        var trimmed = Array(deduped.prefix(Self.maxSlots))
        if trimmed.isEmpty,
           let firstBuiltIn = Self.defaultPinnedURIs.first {
            trimmed = [firstBuiltIn]
        }
        return PinnedAgentGridConfig(
            pinnedURIs: trimmed,
            displayMode: displayMode,
            lastRearrangedAt: lastRearrangedAt
        )
    }

    // MARK: - Mutation helpers

    /// Append `uri` to the end of the grid if there's room and it isn't
    /// already pinned. Returns the mutated config.
    public func pinning(_ uri: String) -> PinnedAgentGridConfig {
        guard !pinnedURIs.contains(uri), pinnedURIs.count < Self.maxSlots else {
            return self
        }
        var copy = self
        copy.pinnedURIs.append(uri)
        copy.lastRearrangedAt = Date()
        return copy
    }

    public func pinningPairedMac(
        _ uri: String,
        pairedMacPrefix: String = "device://paired-mac/"
    ) -> PinnedAgentGridConfig {
        guard uri.hasPrefix(pairedMacPrefix) else {
            return pinning(uri)
        }
        var copy = self
        copy.pinnedURIs.removeAll { $0 == uri }
        copy.pinnedURIs.insert(uri, at: 0)
        copy.pinnedURIs = Array(copy.pinnedURIs.prefix(Self.maxSlots))
        copy.lastRearrangedAt = Date()
        return copy.sanitized()
    }

    /// Remove `uri` from the grid. Returns the mutated config.
    public func unpinning(_ uri: String) -> PinnedAgentGridConfig {
        var copy = self
        copy.pinnedURIs.removeAll { $0 == uri }
        copy.lastRearrangedAt = Date()
        return copy.sanitized()
    }

    /// Move a slot from `sourceIndex` to `destinationIndex`. Returns the
    /// mutated config. Out-of-range indices are clamped.
    public func moving(from sourceIndex: Int, to destinationIndex: Int) -> PinnedAgentGridConfig {
        var copy = self
        let safeFrom = max(0, min(sourceIndex, copy.pinnedURIs.count - 1))
        guard safeFrom < copy.pinnedURIs.count else { return self }
        let value = copy.pinnedURIs.remove(at: safeFrom)
        let safeTo = max(0, min(destinationIndex, copy.pinnedURIs.count))
        copy.pinnedURIs.insert(value, at: safeTo)
        copy.lastRearrangedAt = Date()
        return copy
    }

    // MARK: - JSON convenience

    public func jsonString() -> String {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.sortedKeys]
        encoder.dateEncodingStrategy = .iso8601
        if let data = try? encoder.encode(self),
           let s = String(data: data, encoding: .utf8) {
            return s
        }
        return "{}"
    }

    public static func from(jsonString raw: String) -> PinnedAgentGridConfig {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty, let data = trimmed.data(using: .utf8) else {
            return .default
        }
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return ((try? decoder.decode(PinnedAgentGridConfig.self, from: data))
            ?? .default).sanitized()
    }
}

// MARK: - Persistence keys

extension PinnedAgentGridConfig {
    public static let userDefaultsKey = "square.pinnedGrid.v1"
    public static let androidDataStoreKey = "square.pinned_grid.v1"
}
